File: 백준/Silver/10799. 쇠막대기/쇠막대기.py
import sys

n_list = list(sys.stdin.readline())
stack = []
cnt = 0  # 조각개수
for i in range(len(n_list)):
    if len(stack)==0:
        stack.append(n_list[i])
    else:
        if n_list[i]=="(":
            stack.append(n_list[i])
        else:  # ) 의 경우
            if n_list[i-1]=="(": #괄호가 바로 닫히면
                stack.pop()     #괄호 pop시키고
                cnt+=len(stack) #막대 개수 세리기 ! 막대는 ( 하는 순간 생기기 때문에 (가 append되면 숫자 생김)
            else:               # 괄호가 바로 닫히지 않으면 )) 의 경우이면
                cnt+=1          #막대가 끝나는 점이면 막대 개수+1(막대가 한개 생기는 것)
                stack.pop()

print(cnt)

File: 백준/Silver/10799. 쇠막대기/쇠막대기.java

import java.io.*;
import java.util.*;

public class Main {
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		// 바로인접한 괄호는 레이저
		// rr게 되어있는 괄호는 막대기
		Stack<Character> stack = new Stack<>();
		// 남아있는 (개수만큼 잘린다
		String temp = br.readLine();
		int cnt =0;
		
		for (int i=0; i<temp.length(); i++) {
			if (temp.charAt(i)=='(') {
				stack.push('(');
			}
			else {
				stack.pop(); //그전의 괄호가 열린거면 레이저(사이즈만큼 더해주기
				if(temp.charAt(i-1)=='(') {
					cnt+=stack.size();
				}
				else {
					cnt++; //단순한 닫힌 괄호는 그냥 선이 끝나는 것(+1)
				}
			}
		}
		System.out.println(cnt);

	}

}





